Sources/MongoSwift/BSON/BsonDecoder.swift

```diff
@@ -525,4 +557,127 @@ private extension DecodingError {
+            }
+            self = output
+        }
+    }
+
+    /// Switch through all possible BSON types (a document can contain any) and try recursively decoding the value
+    /// stored under `key` as that type from the provided keyed container.
+    private static func recursivelyDecodeKeyed(key: _BsonKey, container: KeyedDecodingContainer<_BsonKey>) throws -> BsonValue {
+        let k = key.stringValue
+        if let value = try? container.decode(Double.self, forKey: key) {
+            return value
+        } else if let value = try? container.decode(String.self, forKey: key) {
+            return value
+        // this will recursively call Document.init(from: decoder Decoder)
+        } else if let value = try? container.decode(Document.self, forKey: key) {
+            return value
+        } else if var nested = try? container.nestedUnkeyedContainer(forKey: key) {
+            var res = [BsonValue]()
+            while !nested.isAtEnd {
+                res.append(try recursivelyDecodeUnkeyed(container: &nested))
+            }
+            return res
+        } else if let value = try? container.decode(Binary.self, forKey: key) {
+            return value
+        } else if let value = try? container.decode(ObjectId.self, forKey: key) {
+            return value
+        } else if let value = try? container.decode(Bool.self, forKey: key) {
+            return value
+        } else if let value = try? container.decode(Date.self, forKey: key) {
+            return value
+        } else if let value = try? container.decode(RegularExpression.self, forKey: key) {
+            return value
+        } else if let value = try? container.decode(CodeWithScope.self, forKey: key) {
+            return value
+        } else if let value = try? container.decode(Int.self, forKey: key) {
+            return value
+        } else if let value = try? container.decode(Int32.self, forKey: key) {
+            return value
+        } else if let value = try? container.decode(Int64.self, forKey: key) {
+            return value
+        } else if let value = try? container.decode(Decimal128.self, forKey: key) {
+            return value
+        } else if let value = try? container.decode(MinKey.self, forKey: key) {
+            return value
+        } else if let value = try? container.decode(MaxKey.self, forKey: key) {
+            return value
+        } else {
+            throw MongoError.typeError(message: "Encountered a value in an keyed container under key \(k) that could not be decoded to any BSON type")
+        }
+    }
+
+    /// Switch through all possible BSON types (a document can contain any) and try recursively decoding the next value
+    /// as that type from the provided unkeyed container.
+    private static func recursivelyDecodeUnkeyed(container: inout UnkeyedDecodingContainer) throws -> BsonValue {
+        if let value = try? container.decode(Double.self) {
```

Comment:
seems like this can be merged with the `keyed` variant above, by just passing `nil` for the key? Am I missing something?